src/backend/cdb/cdbmutate.c

```diff
@@ -788,6 +809,123 @@ apply_motion_mutator(Node *node, ApplyMotionState *context)
 			return plan_tree_mutator(node, apply_motion_mutator, context);
 	}
 
+	/*
+	 * For UPDATE/DELETE, we check if there's any Redistribute or Broadcast
+	 * Motions before scan in the same subtree for the table we're going to
+	 * modify. If we encounter the scan before any motions, then we can elide
+	 * unneccessary Explicit Redistribute Motion.
+	 */
+	if (IsA(node, ModifyTable))
+	{
+		ModifyTable *mt = (ModifyTable *) node;
+
+		if (mt->operation == CMD_UPDATE || mt->operation == CMD_DELETE)
+		{
+			ListCell   *lcr;
+
+			/*
+			 * Sanity check, since we don't allow multiple ModifyTable nodes
+			 * in the same plan.
+			 */
+			Assert(context->mt.resultRtis == NULL);
+			Assert(context->mt.nMotionsAbove == 0);
+			Assert(!context->mt.needExplicitMotion);
+			Assert(!context->mt.isChecking);
+
+			/*
+			 * When UPDATE/DELETE occurs on a partitioned table, or a table that
+			 * is a part of inheritance tree, ModifyTable node will have more
+			 * than one relation in resultRelations.
+			 *
+			 * We make a list of resulting relations' indexes to compare them
+			 * later.
+			 */
+			foreach(lcr, mt->resultRelations)
+			{
+				Index		rti = lfirst_int(lcr);
+
+				context->mt.resultRtis = bms_add_member(context->mt.resultRtis,
+														rti);
+			}
+
+			context->mt.isChecking = true;
+		}
+	}
+
+	/*
+	 * Elide Explicit Redistribute Motion if there's no motions between the
+	 * scan and the ModifyTable.
+	 */
+	if (context->mt.isChecking)
+	{
+
+		/*
+		 * Remember if we are descending into a Redistribute or Broadcast
+		 * Motion node.
+		 */
+		if (IsA(node, Motion))
+		{
+			Motion	   *motion = (Motion *) node;
+
+			if (motion->motionType == MOTIONTYPE_HASH ||
+				(motion->motionType == MOTIONTYPE_FIXED && motion->isBroadcast))
+			{
+				context->mt.nMotionsAbove += 1;
+			}
+		}
+
+		/*
+		 * If this is a scan and it's scanrelid matches ModifyTable's relid,
+		 * we encountered a scan before any Motions.
+		 */
+		else
+		{
+			switch (nodeTag(node))
+			{
+				case T_SeqScan:
+				case T_DynamicSeqScan:
+				case T_ExternalScan:
+				case T_IndexScan:
+				case T_DynamicIndexScan:
+				case T_IndexOnlyScan:
+				case T_BitmapIndexScan:
+				case T_DynamicBitmapIndexScan:
+				case T_BitmapHeapScan:
+				case T_DynamicBitmapHeapScan:
+				case T_TidScan:
+				case T_SubqueryScan:
+				case T_FunctionScan:
+				case T_TableFunctionScan:
+				case T_ValuesScan:
+				case T_CteScan:
+				case T_WorkTableScan:
+				case T_ForeignScan:
```

Comment:
Are all the types of scans listed here can scan relationships that can be modified? At least of the Scans presented here that cannot scan real relationships is `T_SubqueryScan`, `T_FunctionScan`, `T_ValuesScan` and `T_CteScan`. There may be others. I also think that it is not worth deleting the scan type if this leads to the appearance of "holes", since in the current version the compiler can optimize this for range checking.

Reply:
It seems that ForeignScan also cannot require Explicit Redistribute, since all operations with such tables take place on the coordinator.

Reply:
> all operations with such tables take place on the coordinator

What about gp_toolkit.__gp_log_segment_ext ?

Reply:
The External Table and the Foreign Table are different things

Reply:
Is there anything else that needs to be done?

Reply:
I think that T_ForeignScan also needs to be removed. If you have no objections.

Reply:
I did not find anything that would say that updating foreign tables occurs only from the coordinator. Instead of the normal actions, the `ExecForeign...()` functions are called.

As it looks like, `file_fdw` has no update functions, but `postgres_fdw` seems to support `UPDATE`/`DELETE` without any special restrictions regarding `mpp_execute`.

Reply:
@bandetto Okay It seems that the Update/Delete for Foreign tables will occur on the coordinator, since the ENTRY distribution type is set for them, which leads to a focus on the coordinator.

In the case of "all segments", UPDATE/DELETE does not work due to the absence of the `gp_segment_id` attribute (and most likely there are other restrictions).

I was unable to get an Explicit Redistribute for the Foreign Table without changes in the course (the execution of the plan in this case did not work anyway).
Also, if we leave Explicit Redistribute, then what about external tables?